### PR TITLE
Bundle darknet in docker base images; add logging to ml_api

### LINKED ISSUE #552: [BUG] app hangs on fresh install of recent Jetson Nano images
Summary of discussion from discord where I and another user (TargetCrafter) are both encountering this problem.
 
**Describe the bug**
Running TSD on my jetson nano (4GB) from a fresh installation works for a bit (~hours to a day or two), but eventually it goes unresponsive: octoprint can't contact the server, and I can't access the webpage (:3334). 

Completely removing and reinstalling the repo and re-linking the octoprint plugin gets it working again, I can also sometimes get it to recover with a full down/up via docker-compose.

The issue appears to be a mismatch in CUDA versions between the cuda driver running the model vs the installed runtime, per logs in the `ml_api` container:
```
/usr/bin/python3: [2022-01-02 16:17:44 +0000] [4128] [INFO] Booting worker with pid: 4128
layer     filters    size              input                output
 Try to load cfg: /app/model/model.cfg, weights: /app/model/model.weights, clear = 0
CUDA status Error: file: ./src/cuda.c : () : line: 203 : build time: Oct 15 2019 - 16:00:39
CUDA Error: CUDA driver version is insufficient for CUDA runtime version
   0  Try to set subdivisions=64 in your cfg-file.
```

Replacing usage of the jetsistant docker image in the ml_api base image with the official L4T base image (nvcr.io/nvidia/l4t-base:r32.4.3) produces an error indicating the model expects CUDART v10: https://pastebin.com/z5QcD48S. Attempting to use the repo fork of jetsistant (https://github.com/clearobject/docker-cuda-jetpack) yields the same result. 

I suspect the model must be rebuilt to work with the new version of cuda. I have no idea how to do this, otherwise I would volunteer a PR.

**To Reproduce**
1. Install the [newest image of the jetson nano developer kit](https://developer.nvidia.com/embedded/learn/get-started-jetson-nano-devkit#write)
2. Install and configure TSD as per [the guide](https://github.com/TheSpaghettiDetective/TheSpaghettiDetective/blob/master/docs/jetson_guide.md)
3. Wait a few hours

**Screenshots**

**Hosting environment (please complete the following information):**
 - OS: Jetson Nano 4gb
 - Version: JetPack 4.6
 - Hardware: Jetson nano

```shell
jetson1@jetson1:~$ dpkg-query --show nvidia-l4t-core
nvidia-l4t-core    32.6.1-20210726122000
```

(Corresponding to JetPack 4.6 - seems to be the source of the "driver version" errors in the ML API per https://github.com/TheSpaghettiDetective/TheSpaghettiDetective/issues/515 but they don't seem to prevent it working initially)

**Additional context**

web container logs: https://pastebin.com/QZpV76nu
ML API logs: https://pastebin.com/yhQea6iA
Tasks container logs: https://pastebin.com/XNnpYEMq
Redis logs: https://pastebin.com/DHTPzasb

### PR DESCRIPTION
**Pull request by smartin015**
Closes #552 - this eliminates `ml_api/lib/detection_model.py` in favor of using a bundled build of Darknet, and adds some debug logging for `ml_api` (off by default). I've tested this PR both on the Jetson Nano as well as a linux workstation w/ nvidia GPU by running the `ml_api` container on its own and passing it images via URL.

**IMPORTANT:** this PR currently removes the ability to run TSD without a GPU. If you still want non-gpu runnable versions I can look into it, but I don't know if they're currently used or not and didn't want to spend extra time on it if it wasn't going to be used.

Other details:

* Switched from `jetsistant` aarch64 image to [nvidia-l4t](https://developer.nvidia.com/embedded/learn/tutorials/jetson-container) image which is supported by nvidia directly
* Upgraded x86_64 image from cuda v9 to v11 - necessary to get the latest Darknet to build.
* Introduced a `LOGLEVEL` environment variable which can be uncommented in the docker-compose.yml file to configure's Flask's `app.logger`
* Used `traceback` to print exceptions when Sentry isn't configured (otherwise they're swallowed silently)
* Incremented the base image version tags from `1.1 to `1.2` (this will require pushing new container images for download) 
* To build darknet on the Jetson, I had to edit `/etc/docker/daemon.json` to set the nvidia runtime as default, otherwise certain header files aren't included at build time. This is documented in `model_development.md` but not in `jetson_guide.md` as all the building happens in the base image, which the user *should* be fetching remotely without having to build it.

### CODE REVIEW

**Comment by kennethjiang** [comment]
@smartin015 This looks like an exciting PR! It's great to see the `detection_modle.py` gone!

Most of the changes you made are good. However, we do need to support running this on CPU-only. Most self-hosted TSD server are running on regular PC than on GPU-based machines. Please make changes necessary to enable that.

Also a heads-up I'm super busy right now with the rebranding project (spoiler: we will have a new name!). I will be slow at reviewing/merging this PR, since changes at this level requires a full regression on the training/testing data.

**Comment by smartin015** [comment]
Great, I'll get to work on adding CPU-only support. My first instinct is to fork into two `Dockerfile.base_(x86_64|aarch64)_gpu` images and change the originals to build darknet without GPU support, so they aren't bloated with CUDA / jetpack libs they'll never use.

**Comment by raymondh2** [comment]
It would likely be best to still compile both the CPU and GPU libraries and let them be selected by the setting in the enviorment file. As maintaining two images for GPU and CPU models will complicate user setup.

**Comment by kennethjiang** [comment]
@smartin015 Are you still working on this PR?

**Comment by smartin015** [comment]
Not actively - I'd been leaving it open as @advisoray wanted to try their hand at #552. I pinged them on the issue.

I've finally finished my latest distraction ([peer-to-peer 3d printing over LAN](https://github.com/smartin015/continuousprint/issues/35#issuecomment-1145320824) among other features) so I do have a bit more time now to look into this. But I'll give it a couple days for a response so I'm not doing redundant work.

**Comment by smartin015** [comment]
Ok, latest commit adds support for CPU. Also had to bump NVIDIA images due to https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212772 not applying to the old label.

I've confirmed this works on x86_64 (workstation w/ GTX 1070) and aarch64 (Jetson Nano) both with CPU (no nvidia runtime, HAS_GPU excluded) and with GPU (runtime:nvidia, HAS_GPU: 1 in docker-compose.yml).

**Comment by kennethjiang** [comment]
> Ok, support for CPU is now bundled into the same base docker images. Also had to bump NVIDIA images due to https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212772 not applying to the old label.
> 
> I've confirmed this works on x86_64 (workstation w/ GTX 1070) and aarch64 (Jetson Nano) both with CPU (no nvidia runtime, HAS_GPU excluded) and with GPU (runtime:nvidia, HAS_GPU: 1 in docker-compose.yml).

Not sure why I missed your update. Sorry. Is the PR ready for review/merge now?

**Comment by smartin015** [comment]
Yes - please review/test and let me know if there's any changes needed, but AFAIK it's ready to merge.

**Comment by kennethjiang** [comment]
Just got a chance to test your PR @smartin015. However, I ran into this error:

```
$ docker-compose up --build -d
[+] Building 2.4s (10/10) FINISHED
 => [obico-server_ml_api internal] load build definition from Dockerfile                                                                                                     0.1s
 => => transferring dockerfile: 272B                                                                                                                                         0.0s
 => [obico-server_web internal] load build definition from Dockerfile                                                                                                        0.1s
 => => transferring dockerfile: 176B                                                                                                                                         0.0s
 => [obico-server_tasks internal] load build definition from Dockerfile                                                                                                      0.0s
 => => transferring dockerfile: 176B                                                                                                                                         0.0s
 => [obico-server_ml_api internal] load .dockerignore                                                                                                                        0.0s
 => => transferring context: 2B                                                                                                                                              0.0s
 => [obico-server_web internal] load .dockerignore                                                                                                                           0.0s
 => => transferring context: 118B                                                                                                                                            0.0s
 => [obico-server_tasks internal] load .dockerignore                                                                                                                         0.0s
 => => transferring context: 118B                                                                                                                                            0.0s
 => ERROR [obico-server_ml_api internal] load metadata for docker.io/thespaghettidetective/ml_api:base-1.2                                                                   1.8s
 => CANCELED [obico-server_web internal] load metadata for docker.io/thespaghettidetective/web:base-1.8                                                                      1.7s
 => [auth] thespaghettidetective/web:pull token for registry-1.docker.io                                                                                                     0.0s
 => [auth] thespaghettidetective/ml_api:pull token for registry-1.docker.io                                                                                                  0.0s
------
 > [obico-server_ml_api internal] load metadata for docker.io/thespaghettidetective/ml_api:base-1.2:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: docker.io/thespaghettidetective/ml_api:base-1.2: not found
```

**Comment by kennethjiang** [comment]
Ah it looks like I need to update the base image first. 

Did you follow `ml_api_base_image_manifest.sh` to build the base? I just want to make sure I accurately reproduce what you have done.

**Comment by smartin015** [comment]
> Ah it looks like I need to update the base image first.

Yep - this PR includes an update to the `FROM` of the base image, as newer versions of darknet didn't run on the older version of CUDA used previously. 

> Did you follow `ml_api_base_image_manifest.sh` to build the base? I just want to make sure I accurately reproduce what you have done.

Sorry about that - I totally overlooked that script. After a brief look, I'm still not sure how I'd incorporate it as it references `build_dockerfile_ml_api_base_arm64v8.sh` and `build_dockerfile_ml_api_base_amd64.sh`, neither of which exist AFAIK.

Here's what I did to create a running container on a jetson nano:

```
cd path/to/repo/ml_api
docker build --tag thespaghettidetective/ml_api:base-1.2 -f Dockerfile.base_aarch64 .
docker-compose build ml_api
docker-compose run --service-ports ml_api
```

On my workstation, the process was the same except `-f Dockerfile.base_aarch64` becomes `-f Dockerfile.base_x86_64`

**Comment by kennethjiang** [comment]
> > Ah it looks like I need to update the base image first.
> 
> Yep - this PR includes an update to the `FROM` of the base image, as newer versions of darknet didn't run on the older version of CUDA used previously.
> 
> > Did you follow `ml_api_base_image_manifest.sh` to build the base? I just want to make sure I accurately reproduce what you have done.
> 
> Sorry about that - I totally overlooked that script. After a brief look, I'm still not sure how I'd incorporate it as it references `build_dockerfile_ml_api_base_arm64v8.sh` and `build_dockerfile_ml_api_base_amd64.sh`, neither of which exist AFAIK.
> 
> Here's what I did to create a running container on a jetson nano:
> 
> ```
> cd path/to/repo/ml_api
> docker build --tag thespaghettidetective/ml_api:base-1.2 -f Dockerfile.base_aarch64 .
> docker-compose build ml_api
> docker-compose run --service-ports ml_api
> ```
> 
> On my workstation, the process was the same except `-f Dockerfile.base_aarch64` becomes `-f Dockerfile.base_x86_64`

Got it. Busy with preparing for MRRF these days. But will get to it soon after that.

**Comment by kennethjiang** [comment]
> `build_dockerfile_ml_api_base_arm64v8.sh` and `build_dockerfile_ml_api_base_amd64.sh`,

build_dockerfile_ml_api_base_arm64v8.sh and build_dockerfile_ml_api_base_amd64.sh are in the "ml_api/scripts" folder.

Do you mean you didn't create a manifest for the "ml_api:base-1.2" image you created? If so, how did you use the same image for both amd and arm architectures?

The commadn `docker build --tag thespaghettidetective/ml_api:base-1.2 -f Dockerfile.base_aarch64 .` will build ml_api:base-1.2 only for the arm arch. My understanding is the users on PCs will fail to run it, right? Have you tested it on an intel-based computer? My knowledge about docker's multi-arch image may be out of date so I could be wrong here.

**Comment by smartin015** [comment]
Ah - `./ml_api/scripts/`, not `./scripts/`. That explains why I didn't see them :)

WRT manifests and cross-arch: I'd never touched this aspect of docker before. I just built the image I needed on the device I wanted to run on. I see what you're doing now, though. 

TL;DR of the notes below: you should be able to

1. run the `build_dockerfile_ml_api_base_*.sh` scripts on the appropriate devices
2. run `ml_api_base_image_manifest.sh` to generate and deploy the manifest / multi-arch image
3. run `docker-compose build ml_api` to use the deployed image from (2)
4. run `docker-compose --service-ports ml_api` and verify the model correctly loads for various configs, on various devices.

The only reason my notes are more complex is because I don't have the necessary credentials to deploy to `docker.io/thespaghettidetective` and had to tweak things to use my own private registry. But I did managed to build a multi-arch image, pull it, and run it on my workstation with no issues.

## notes

So, running on workstation:

```
cd ml_api && ./scripts/build_dockerfile_ml_api_base_amd64.sh 1.2
```

Builds the image, but fails to push as I lack permissions to actually deploy the image. I had to fudge this using my own private registry:

```
docker tag thespaghettidetective/ml_api:base-1.2-linux-amd64 <redacted>/ml_api:base-1.2-linux-amd64
docker push <redacted>/ml_api:base-1.2-linux-amd64
``` 

Had to do the same with the Jetson:

```
cd ml_api && ./scripts/build_dockerfile_ml_api_base_arm64v8.sh 1.2
docker tag thespaghettidetective/ml_api:base-1.2-linux-arm64v8 <redacted>/ml_api:base-1.2-linux-arm64v8
docker push <redacted>/ml_api:base-1.2-linux-arm64v8
```

And of course this required changes as well to `ml_api_base_image_manifest.sh`, replacing `thespaghettidetective` with my private registry (and also enable execution permissions). `docker pull` / `docker inspect` seems to indicate the correct version of the image is getting pulled, and (once more replacing the registry in ml_api/Dockerfile) running `docker-compose build ml_api` and `docker-compose run --service-ports ml_api` Brings the server up and running as expected.

**Comment by kennethjiang** [comment]
@smartin015 Thank you for the details. Now I'm back from MRRF. Should be able to spend more time on this PR.

Building the base image worked with a glitch on my Mac. But when I tried to do the same on the Jetson Nano, I got this error:

g++ -std=c++11 -std=c++11 -Iinclude/ -I3rdparty/stb/include -DGPU -I/usr/local/cuda/include/ -Wall -Wfatal-errors -Wno-unused-result -Wno-unknown-pragmas -fPIC -Ofast -DGPU -fPIC -c ./src/image_opencv.cpp -o obj/image_opencv.o
In file included from ./src/image.h:3:0,
                 from ./src/image_opencv.h:4,
                 from ./src/image_opencv.cpp:1:
include/darknet.h:43:10: fatal error: cublas_v2.h: No such file or directory
 #include <cublas_v2.h>
          ^~~~~~~~~~~~~

[Console output attached](https://github.com/TheSpaghettiDetective/obico-server/files/8995570/build.stdout.zip).

**Comment by kennethjiang** [comment]
@smartin015 In case you missed my previous comment.

**Comment by smartin015** [comment]
Sorry for the delay, been a busy week - I'll take a look into the build error during the weekend.

**Comment by smartin015** [comment]
Okay - finally able to sit down and take a look. The `-DGPU -I/usr/local/cuda/include/` makes me suspect this is an nvidia runtime issue, as builds normally use the default docker runtime which doesn't mount all the necessary host cuda libraries into the container. 

Could you take a look at your docker daemon config? I suspect you need to add `"default-runtime": "nvidia",` to the json in order to invoke nvidia runtime on build (alas, there's no build argument to specify the runtime to use). Here's my config:

```
sudo cat /etc/docker/daemon.json
...
{
    "default-runtime": "nvidia",
    "runtimes": {
        "nvidia": {
            "path": "nvidia-container-runtime",
            "runtimeArgs": []
        }
    }
}
```

Then run `sudo systemctl restart docker`  (or reboot) to restart the docker engine, and retry your build command.

Just as a sanity check, I ran the same command as you, just with `--no-cache`. 

```
sudo docker build --no-cache . -f Dockerfile.base_aarch64 -t thespaghettidetective/ml_api:base-1.2-linux-arm64v8
```

And it worked without a problem.

My jetson version:

```
cat /etc/nv_tegra_release
# R32 (release), REVISION: 6.1, GCID: 27863751, BOARD: t210ref, EABI: aarch64, DATE: Mon Jul 26 19:20:30 UTC 2021
```

**Comment by kennethjiang** [comment]
Sorry for the delay! My ` /etc/docker/daemon.json` is the same as yours. But this time it worked! I guess `--no-cache` made a difference? Not sure...

Moving on to the next steps.

**Comment by kennethjiang** [comment]
@smartin015 Let me know if you still want to work with me to merge this PR.

**Comment by smartin015** [comment]
Sorry for the delay - I didn't see your edit until just now. 

I'm currently away and unable to reach my jetson, but I should be able to push an image a few days from now when I get back.

**Comment by kennethjiang** [comment]
No problem. Moving forward I will make sure I tag you so that you will get notified.

**Comment by kennethjiang** [comment]
@smartin015 It looks like you didn't get a chance to work on it. I'm closing this PR since it's not a good idea to have a long standing PR (it's probably difficult to merge anyway). But feel free to reopen it when you get a chance to finish it up. Thank you for your effort on this complex PR and I'm sorry I can't merge it.

**Comment by smartin015** [comment]
Yep, apologies for falling off the wagon here, I've just been way too wrapped up in other projects and it hasn't meshed well with a slow-running PR. 

Hopefully some other enterprising contributor can take what's left of this and finish it off.
